java/cdk/src/main/java/sleeper/cdk/stack/QueryStack.java

```diff
@@ -128,71 +133,191 @@ public QueryStack(Construct scope,
+                instanceProperties.get(ID).toLowerCase(Locale.ROOT), "query-register"));
+
+        String leafQueryFunctionName = Utils.truncateTo64Characters(String.join("-", "sleeper",
+                instanceProperties.get(ID).toLowerCase(Locale.ROOT), "query-executor"));
+
+        // Lambda to register queries and post leaf partition queries to a queue
+        IFunction queryRegisterLambda = createFunction("QueryRegisterLambda", queryJar, instanceProperties, registerFunctionName,
+            "sleeper.query.lambda.SqsQueryRegisterLambda::handleRequest",
+                "When a query arrives on the query SQS queue, this lambda is invoked to look for leaf partition queries");
+
+        // Lambda to execute the queries
+        IFunction leafPartitionQueryLambda = createFunction("QueryExecutorLambda", queryJar, instanceProperties, leafQueryFunctionName,
+                "sleeper.query.lambda.SqsLeafPartitionQueryLambda::handleRequest",
+                "When a query arrives on the query SQS queue, this lambda is invoked to execute the query");
+
+        // Add the queue as a source of events for the lambdas
+        SqsEventSourceProps eventSourceProps = SqsEventSourceProps.builder()
+                .batchSize(1)
+                .build();
+
+        queryRegisterLambda.addEventSource(new SqsEventSource(queryRegisterQueue, eventSourceProps));
+        leafPartitionQueryLambda.addEventSource(new SqsEventSource(leafPartitionQueriesQueue, eventSourceProps));
+
+        /* Grant the lambdas permission to read from the Dynamo tables, read from
+           the S3 bucket, write back to the query queue and write to the results
+           queue and S3 bucket */
+        setPermissionsForLambda(coreStacks, jarsBucket, queryRegisterLambda, queryTrackingTable, queryRegisterQueue);
+        setPermissionsForLambda(coreStacks, jarsBucket, leafPartitionQueryLambda, queryTrackingTable, queryRegisterQueue, queryResultsQueue, queryResultsBucket);
+
+        /* Attach a policy to allow the lambda to put results in any S3 bucket or on any SQS queue.
+           These policies look too open, but it's the only way to allow clients to be able to write to their
+           buckets and queues.*/
+        PolicyStatementProps policyStatementProps = PolicyStatementProps.builder()
+                .effect(Effect.ALLOW)
+                .actions(Arrays.asList(S3Actions.PutObject.getActionName(), SQSActions.SendMessage.getActionName()))
+                .resources(Collections.singletonList("*"))
+                .build();
+        PolicyStatement policyStatement = new PolicyStatement(policyStatementProps);
+        Policy policy = new Policy(this, "PutToAnyS3BucketAndSendToAnySQSPolicy");
+        policy.addStatements(policyStatement);
+        Objects.requireNonNull(queryRegisterLambda.getRole()).attachInlinePolicy(policy);
+        Objects.requireNonNull(leafPartitionQueryLambda.getRole()).attachInlinePolicy(policy);
```

Comment:
You could extract this policy attaching code to a new function, something like `attachPolicy(IFunction lambda)`, and convert the comment above into javadoc for the new function.

Reply:
Class has been refactored